pkg/cafs/freelists.go

```diff
@@ -1,63 +1,362 @@
 package cafs
 
 import (
+	"math"
 	"sync"
+	"unsafe"
+
+	"github.com/docker/go-units"
+)
+
+const (
+	leaf1MB uint32 = 1 * units.MiB
+	leaf2MB uint32 = 2 * units.MiB
+	leaf3MB uint32 = 3 * units.MiB
+	leaf4MB uint32 = 4 * units.MiB
 )
 
-/* this is a memory pooling technique adapted from
- * github.com/jacobsa/fuse/internal/freelist.
- */
+// BytesToBuffers returns the number of buffers for a given
+// size target in bytes, given a leaf size. Buffers are rounded up,
+// so we may exceed the target by at most the size of a leaf.
+func BytesToBuffers(target int, leafSize uint32) int {
+	return int(math.Ceil(float64(target) / float64(leafSize)))
+}
+
+// FreeList knows how to allocate buffers and release them
+type FreeList interface {
```

Comment:
vis one idea per commit:  i wouldn't mind seeing this freelist re-write as a separate commit.

it's not clear to me if we actually need all the complexity encoded here, yet i do want some of the improvements, and i definitely want to land the majority of this patch as a whole.

Reply:
afterthought

note on terminology:  freelist, alloc/free, and buffer pool are all words that we use to refer to approximately the same kind of memory management setup
* alloc/free - system-level memory management, either via standard C library or Go runtime internals.
  while here, note that it's also not altogther uncommon to write custom `alloc`/`free` routines on systems (e.g. plan9) that weren't written expressly to optimize memory management and that haven't been productionalized to the extent that linux, etc. have
* freelist -  `github.com/jacobsa/fuse/internal/freelist` is my original source for hearing about this.  having read some more about them, i take freelist to connote pretty much strictly a slice of some fixed type that's used to recycle buffers in the manner of the original implementation from `jacobsa/fuse`
* buffer pool - Ritesh's mention of this when i was looking into the memory usage was the first time i heard about it.  this is the usual term for golang libraries responsible for allocating and releasing buffers of a more arbitrary nature of a more arbitrary means

so i guess all i'd suggest doing here for now is renaming to buffer pool, and spinning this comment off into a separate issue to look into using a drop-in replacement later.

in the original implementation, i recall making the decision between using freelists and buffer pools with the intent of evolving the program by choosing a golang buffer pool library.  like, essentially i threw the freelist in place b/c i knew it solved the immediate error and didn't want to introduce an upstream dep without some investigation.

in the long run, as we discussed on the most recent call, i suspect (or perhaps hope is a better term) we'd be better off leveraging the [golang/linux] ecosystem than getting into the nitty-gritty of this kind of performance tuning as part of the datamon project, at least for this version.  like, there is a semi-serious thread of conversation about rust re-writes and such.

----

again, quite enthused about this patch.

Reply:
> "one idea by commit" 

I generally agree with this principle and you know I have been trying to stick to this as much as I could.

Regarding this particular PR, however,  we are in a situation in which the change is driven by a user report such as "reading a big file on fuse RO mount is slow".

Besides, previous discussions regarding performances of the fuse mount led to trying to work on the memory footprint.

Addressing this issue required:
(i) to understand the whole chain of communication across several packages, for this specific use case -- that makes a lot of code, but this is really only covering the read of file on a RO fuse mount, -- improving readability was really required
(ii) to figure out many smaller issues, which were all contributing somehow to the degradation of performances, or to a degraded understanding of how this works - essentially this has driven toward injecting more debug logging everywhere (i'd like to revert on that later on, favouring  exported metrics  over debug traces),
(iii) to unroll a nasty yarn of fixes, as applying smaller fixes did in turn uncover hidden or latent bugs,
(iv) introducing/re-enacting ways to test things

Reply:
> in the original implementation, i recall making the decision between using freelists and buffer pools with the intent of evolving the program by choosing a golang buffer pool library. like, essentially i threw the freelist in place b/c i knew it solved the immediate error and didn't want to introduce an upstream dep without some investigation

I've tried to fix and improve the simplistic approach to freelists and cache buffers, without challenging the very existence of this feature.
Managing memory and caching FS is essentially a complex thing, which **cannot** be addressed by simple approaches.
I don't want to keep this in this state, quite the contrary.

Rather than finding a "pool library" I am more in favor of selecting a full-fledged embedded KV store to which we could delegate _all_ the internal caching and relieve pressure on memory (and memory management code...). My current pick is badgerdb. You may share your thoughts about this idea. That would be a drop-in replacement of the caching/pooling logics, but not the prefetching one.

Of course, the fuse cache tuning is also a possibility, that I would think as complementary.

Reply:
My next move on this is to conduct some research on supporting an alternate way of caching things, using badgerdb as KV store on local FS. My intention is to let the current mechanism as a viable alternative, then eventually drop it if the KV approach proves better in all use cases.

Reply:
follow-up note: open discussion thread

Reply:
> selecting a full-fledged embedded KV store to which we could delegate all the internal caching and relieve pressure on memory (and memory management code...). My current pick is badgerdb.

> You may share your thoughts about this idea.

so much the better.  i mean, i know we definitely don't want to write our own caches and memory allocators.

the stronger opinion here to recycle badgerdb to replace both the cache _and_ memory allocator is all the better by me.

i'll leave this thread of discussion for now without spinning off additional issues.  thanks for the additional note on intent.